lib/record_store/provider/oracle_cloud_dns.rb

```diff
@@ -0,0 +1,182 @@
+require 'oci'
+
+module RecordStore
+  class Provider::OracleCloudDNS < Provider
+    class Error < StandardError; end
+
+    class << self
+      def client
+        @client ||= begin
+          config = OCI::Config.new
+          config.user = secrets['user']
+          config.fingerprint = secrets['fingerprint']
+          config.key_content = secrets['key_content']
+          config.tenancy = secrets['tenancy']
+          config.region = secrets['region']
+          config.validate
+          OCI::Dns::DnsClient.new(config: config)
+        end
+      end
+
+      # Downloads all the records from the provider.
+      #
+      # Returns: an array of `Record` for each record in the provider's zone
+      def retrieve_current_records(zone:, stdout: $stdout)
+        client.get_zone_records(zone)
+          .flat_map { |response| response.data.items }
+          .flat_map { |api_record| build_from_api(api_record) }
+          .compact
+      rescue OCI::Errors::HttpRequestBasedError
+        stdout.puts "Cannot build record for zone: #{zone}"
+        raise
+      end
+
+      # Returns an array of the zones managed by provider as strings
+      def zones
+        client.list_zones(secrets['compartment_id']).data.map(&:name)
+      end
+
+      private
+
+      # Creates a new record to the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # record - a kind of `Record`
+      def add(record, zone)
+        patch_add_record = [
+          OCI::Dns::Models::RecordOperation.new(
+            domain: record.fqdn,
+            rtype: record.type,
+            ttl: record.ttl,
+            rdata: record.rdata_txt,
+            operation: 'ADD',
+          ),
+        ]
+
+        client.patch_zone_records(
+          zone,
+          OCI::Dns::Models::PatchZoneRecordsDetails.new(items: patch_add_record)
+        )
+      end
+
+      # Deletes an existing record from the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # record - a kind of `Record`
+      def remove(record, zone)
+        record_fqdn = Record.ensure_ends_without_dot(record.fqdn)
+        found_record = client.get_zone_records(
+          zone,
+          rtype: record.type,
+          zone_version: client.get_zone(zone).data.version,
```

Comment:
Is it necessary to lock the retrieval of records down to a `zone_version`?  It doesn't look as if the `zone_version` is used during the actual PATCH, so just wondering.  This mirrors a comment @XaF had earlier elsewhere regarding `zone_version` with `get_zone_records`.

Reply:
This can be removed.